File: kotlin-archive/jam-pvm/src/main/kotlin/io/forge/jam/pvm/program/ProgramBlob.kt
package io.forge.jam.pvm.program

import io.forge.jam.pvm.PvmLogger
import io.forge.jam.pvm.engine.InstructionSet
import io.forge.jam.pvm.engine.InstructionSetKind
import io.forge.jam.pvm.engine.JumpTable
import io.forge.jam.pvm.engine.RuntimeInstructionSet

/**
 * A partially deserialized PolkaVM program.
 */
data class ProgramBlob(
    var isaKind: InstructionSetKind? = null,  // Track the ISA kind for proper instruction validation
    var is64Bit: Boolean = false,
    var roDataSize: UInt = 0u,
    var rwDataSize: UInt = 0u,
    var actualRwDataLen: UInt = 0u,  // The actual rwData content length (without heap pages)
    var stackSize: UInt = 0u,
    var heapPages: UInt = 0u,  // Number of heap empty pages from program blob
    var roData: ArcBytes = ArcBytes.empty(),
    var rwData: ArcBytes = ArcBytes.empty(),
    var code: ArcBytes = ArcBytes.empty(),
    var jumpTable: ArcBytes = ArcBytes.empty(),
    var jumpTableEntrySize: Byte = 0,
    var bitmask: ArcBytes = ArcBytes.empty(),
    private var importOffsets: ArcBytes = ArcBytes.empty(),
    private var importSymbols: ArcBytes = ArcBytes.empty(),
    private var exports: ArcBytes = ArcBytes.empty(),
    private var debugStrings: ArcBytes = ArcBytes.empty(),
    private var debugLineProgramRanges: ArcBytes = ArcBytes.empty(),
    private var debugLinePrograms: ArcBytes = ArcBytes.empty()
) {

    fun jumpTable(): JumpTable {
        return JumpTable(jumpTable, jumpTableEntrySize.toUInt())
    }

    fun <I : InstructionSet> instructionsBoundedAt(instructionSet: I, offset: ProgramCounter): Instructions<I> {
        return Instructions.new(
            code = code.toByteArray(),
            bitmask = bitmask.toByteArray(),
            offset = offset.value,
            instructionSet = instructionSet,
            isBounded = true
        )
    }

    fun isJumpTargetValid(instructionSet: RuntimeInstructionSet, offset: ProgramCounter): Boolean {
        return Program.isJumpTargetValid(instructionSet, code.toByteArray(), bitmask.asRef(), offset.value)
    }

    /**
     * Get the instruction set for this program blob.
     * Uses the ISA kind if available, otherwise falls back to legacy is64Bit-based selection.
     */
    fun instructionSet(allowSbrk: Boolean = true): RuntimeInstructionSet =
        isaKind?.let { RuntimeInstructionSet.fromKind(it, allowSbrk) }
            ?: RuntimeInstructionSet(allowSbrk, is64Bit)

    companion object {
        private val logger = PvmLogger(ProgramBlob::class.java)
        private val VM_MAXIMUM_JUMP_TABLE_ENTRIES: UInt = 16u * 1024u * 1024u
        private val VM_MAXIMUM_CODE_SIZE: UInt = 32u * 1024u * 1024u
        private const val VERSION_DEBUG_LINE_PROGRAM_V1: Byte = 1

        fun fromParts(parts: ProgramParts): Result<ProgramBlob> = runCatching {
            val blob = ProgramBlob(
                isaKind = parts.isaKind,
                is64Bit = parts.is64Bit,
                roDataSize = parts.roDataSize,
                rwDataSize = parts.rwDataSize,
                actualRwDataLen = parts.actualRwDataLen,
                stackSize = parts.stackSize,
                heapPages = parts.heapPages,
                roData = parts.roData,
                rwData = parts.rwData,
                exports = parts.exports,
                importSymbols = parts.importSymbols,
                importOffsets = parts.importOffsets,
                debugStrings = parts.debugStrings,
                debugLineProgramRanges = parts.debugLineProgramRanges,
                debugLinePrograms = parts.debugLinePrograms
            )
            if (blob.roData.asRef().size > blob.roDataSize.toInt()) {
                throw ProgramParseError(
                    ProgramParseErrorKind.Other("size of the read-only data payload exceeds the declared size of the section")
                )
            }

            if (blob.rwData.asRef().size > blob.rwDataSize.toInt()) {
                throw ProgramParseError(
                    ProgramParseErrorKind.Other("size of the read-write data payload exceeds the declared size of the section")
                )
            }

            if (parts.codeAndJumpTable.asRef().isEmpty()) {
                throw ProgramParseError(ProgramParseErrorKind.Other("no code found"))
            }

            // Use JAM-specific parsing for JamV1 ISA
            if (parts.isaKind == InstructionSetKind.JamV1) {
                parseJamCodeSection(parts, blob)
            } else {
                parsePolkavmCodeSection(parts, blob)
            }

            blob
        }

        /**
         * Parse JAM format code section.
         */
        private fun parseJamCodeSection(parts: ProgramParts, blob: ProgramBlob) {
            val data = parts.codeAndJumpTable.toByteArray()

            if (data.isEmpty()) {
                throw ProgramParseError(ProgramParseErrorKind.Other("JAM code section is empty"))
            }

            var offset = 0

            // Read jump table entries count
            val (jumpTableEntryCount, offset1) = decodeJamNatural(data, offset)
            offset = offset1

            if (offset >= data.size) {
                throw ProgramParseError(ProgramParseErrorKind.Other("JAM code section truncated: missing encodeSize"))
            }

            // Read jump table entry size (single byte)
            val jumpTableEntrySize = (data[offset].toInt() and 0xFF).toByte()
            offset++

            // Read code length
            val (codeLength, offset2) = decodeJamNatural(data, offset)
            offset = offset2

            // Calculate jump table length
            val jumpTableLength = jumpTableEntryCount.toInt() * jumpTableEntrySize.toInt()

            // Read jump table
            val jumpTableEnd = offset + jumpTableLength
            if (jumpTableEnd > data.size) {
                throw ProgramParseError(
                    ProgramParseErrorKind.Other("JAM code section truncated: jumpTable extends beyond end")
                )
            }
            blob.jumpTableEntrySize = jumpTableEntrySize
            blob.jumpTable = ArcBytes.fromStatic(data.copyOfRange(offset, jumpTableEnd))
            offset = jumpTableEnd

            // Read code
            val codeEnd = offset + codeLength.toInt()
            if (codeEnd > data.size) {
                throw ProgramParseError(
                    ProgramParseErrorKind.Other("JAM code section truncated: code extends beyond end (offset=$offset, codeLength=$codeLength, dataSize=${data.size})")
                )
            }
            blob.code = ArcBytes.fromStatic(data.copyOfRange(offset, codeEnd))
            offset = codeEnd

            // Read bitmask (remaining bytes)
            val bitmaskLength = data.size - offset
            blob.bitmask = ArcBytes.fromStatic(data.copyOfRange(offset, data.size))

            // Validate bitmask length
            var expectedBitmaskLength = codeLength.toInt() / 8
            val isBitmaskPadded = codeLength.toInt() % 8 != 0
            if (isBitmaskPadded) expectedBitmaskLength++

            if (bitmaskLength != expectedBitmaskLength) {
                throw ProgramParseError(
                    ProgramParseErrorKind.Other(
                        "JAM bitmask length mismatch: expected $expectedBitmaskLength, got $bitmaskLength (codeLength=$codeLength)"
                    )
                )
            }
        }

        /**
         * Decode a JAM natural number
         * Returns the decoded value and the new offset.
         */
        private fun decodeJamNatural(data: ByteArray, offset: Int): Pair<Long, Int> {
            if (offset >= data.size) {
                throw ProgramParseError(ProgramParseErrorKind.Other("JAM natural number: unexpected end of data"))
            }

            val firstByte = data[offset].toInt() and 0xFF
            if (firstByte == 0) {
                return Pair(0L, offset + 1)
            }

            // Count leading zeros of inverted byte to get number of additional bytes
            val inverted = firstByte.inv() and 0xFF
            var byteLength = 0
            for (i in 0 until 8) {
                if ((inverted and (0x80 shr i)) != 0) break
                byteLength++
            }

            if (offset + 1 + byteLength > data.size) {
                throw ProgramParseError(ProgramParseErrorKind.Other("JAM natural number: truncated"))
            }

            // Read additional bytes (little-endian)
            var res: Long = 0
            for (i in 0 until byteLength) {
                res = res or ((data[offset + 1 + i].toLong() and 0xFF) shl (8 * i))
            }

            // Mask for top bits from first byte
            val mask = (1 shl (8 - byteLength)) - 1
            val topBits = firstByte and mask

            val value = res + (topBits.toLong() shl (8 * byteLength))
            return Pair(value, offset + 1 + byteLength)
        }

        /**
         * Parse standard PolkaVM format code section.
         */
        private fun parsePolkavmCodeSection(parts: ProgramParts, blob: ProgramBlob) {
            val reader = ArcBytesReader(parts.codeAndJumpTable)
            val initialPosition = reader.position

            val jumpTableEntryCount = reader.readVarintInternal().getOrThrow()
            if (jumpTableEntryCount > VM_MAXIMUM_JUMP_TABLE_ENTRIES) {
                throw ProgramParseError(ProgramParseErrorKind.Other("the jump table section is too long"))
            }

            val jumpTableEntrySize = reader.readByte().getOrThrow()
            if (jumpTableEntrySize !in 0..4) {
                throw ProgramParseError(ProgramParseErrorKind.Other("invalid jump table entry size"))
            }

            val codeLength = reader.readVarintInternal().getOrThrow()
            if (codeLength > VM_MAXIMUM_CODE_SIZE) {
                throw ProgramParseError(ProgramParseErrorKind.Other("the code section is too long"))
            }

            val jumpTableLength = jumpTableEntryCount.toInt() * jumpTableEntrySize.toInt()
            blob.jumpTableEntrySize = jumpTableEntrySize
            blob.jumpTable = reader.readSliceAsBytes(jumpTableLength).getOrThrow()
            blob.code = reader.readSliceAsBytes(codeLength.toInt()).getOrThrow()

            val bitmaskLength = parts.codeAndJumpTable.asRef().size - (reader.position - initialPosition)
            blob.bitmask = reader.readSliceAsBytes(bitmaskLength).getOrThrow()

            var expectedBitmaskLength = blob.code.asRef().size / 8
            val isBitmaskPadded = blob.code.asRef().size % 8 != 0
            if (isBitmaskPadded) expectedBitmaskLength++

            if (blob.bitmask.asRef().size != expectedBitmaskLength) {
                throw ProgramParseError(
                    ProgramParseErrorKind.Other("the bitmask length doesn't match the code length")
                )
            }

            if (isBitmaskPadded) {
                val lastByte = blob.bitmask.asRef().last()
                val paddingBits = blob.bitmask.asRef().size * 8 - blob.code.asRef().size
                val paddingMask = (0b10000000.toByte().toInt() shr (paddingBits - 1)).toByte()

                if ((lastByte.toInt() and paddingMask.toInt()) != 0) {
                    throw ProgramParseError(
                        ProgramParseErrorKind.Other("the bitmask is padded with non-zero bits")
                    )
                }
            }
        }
    }
}

File: modules/pvm/src/main/scala/io/forge/jam/pvm/program/ProgramBlob.scala
package io.forge.jam.pvm.program

import io.forge.jam.pvm.Abi

/**
 * Parsed PVM program blob.
 *
 * Contains all the components extracted from a PVM binary:
 * - Code section
 * - Bitmask for instruction boundaries
 * - Jump table
 * - Read-only and read-write data
 * - Program mode (32-bit or 64-bit)
 *
 * @param code The executable code section
 * @param bitmask Instruction boundary bitmask
 * @param jumpTable The jump table for indirect jumps
 * @param is64Bit True if this is a 64-bit program
 * @param roData Read-only data section
 * @param rwData Read-write data section
 * @param stackSize Initial stack size
 */
final case class ProgramBlob(
  code: Array[Byte],
  bitmask: Array[Byte],
  jumpTable: JumpTable,
  is64Bit: Boolean,
  roData: Array[Byte],
  rwData: Array[Byte],
  stackSize: Int,
  heapPages: Int = 0,
  originalRwDataLen: Int = -1
):
  /** Check if code size is within limits */
  def isCodeSizeValid: Boolean =
    code.length <= Abi.VmMaximumCodeSize.signed

  /** Check if jump table size is within limits */
  def isJumpTableSizeValid: Boolean =
    jumpTable.len <= Abi.VmMaximumJumpTableEntries.signed

  /** Total code length */
  def codeLength: Int = code.length

object ProgramBlob:

  /**
   * Creates a ProgramBlob from raw code+jumptable bytes (JAM format).
   *
   * - Varint: jump table entry count
   * - Byte: jump table entry size
   * - Varint: code length
   * - Jump table bytes
   * - Code bytes
   * - Bitmask bytes
   */
  def fromCodeAndJumpTable(
    data: Array[Byte],
    roData: Array[Byte] = Array.empty,
    rwData: Array[Byte] = Array.empty,
    stackSize: Int = 0,
    is64Bit: Boolean = true,
    heapPages: Int = 0,
    originalRwDataLen: Int = -1
  ): Option[ProgramBlob] =
    if data.isEmpty then return None

    var offset = 0

    // Read jump table entry count (varint)
    val jumpTableEntryCount = readTestVarint(data, offset) match
      case Some((count, newOffset)) =>
        offset = newOffset
        count.toInt
      case None => return None

    if offset >= data.length then return None

    // Read jump table entry size (single byte)
    val jumpTableEntrySize = data(offset) & 0xFF
    offset += 1

    // Read code length (varint)
    val codeLength = readTestVarint(data, offset) match
      case Some((len, newOffset)) =>
        offset = newOffset
        len.toInt
      case None => return None

    // Calculate jump table length
    val jumpTableLength = jumpTableEntryCount * jumpTableEntrySize

    // Calculate bitmask length
    val bitmaskLength = (codeLength + 7) / 8

    // Verify we have enough data
    val expectedTotal = offset + jumpTableLength + codeLength + bitmaskLength
    if data.length < expectedTotal then return None

    // JAM format: jump table comes BEFORE code
    // Extract jump table
    val jumpTableData = new Array[Byte](jumpTableLength)
    if jumpTableLength > 0 then
      System.arraycopy(data, offset, jumpTableData, 0, jumpTableLength)
    offset += jumpTableLength

    // Extract code
    val code = new Array[Byte](codeLength)
    System.arraycopy(data, offset, code, 0, codeLength)
    offset += codeLength

    // Extract bitmask
    val bitmask = new Array[Byte](bitmaskLength)
    System.arraycopy(data, offset, bitmask, 0, bitmaskLength)

    val jumpTable = JumpTable(jumpTableData, jumpTableEntrySize)

    Some(ProgramBlob(
      code = code,
      bitmask = bitmask,
      jumpTable = jumpTable,
      is64Bit = is64Bit,
      roData = roData,
      rwData = rwData,
      stackSize = stackSize,
      heapPages = heapPages,
      originalRwDataLen = originalRwDataLen
    ))

  /**
   * Read a varint from data at offset. Used for test vectors.
   * Returns (value, newOffset) or None on failure.
   */
  private def readTestVarint(data: Array[Byte], offset: Int): Option[(Long, Int)] =
    if offset >= data.length then return None

    val firstByte = data(offset) & 0xFF

    // Count leading 1 bits to determine length
    var byteLength = 0
    var temp = firstByte
    while (temp & 0x80) != 0 && byteLength < 8 do
      byteLength += 1
      temp = (temp << 1) & 0xFF

    if offset + 1 + byteLength > data.length then return None

    // Read additional bytes (little-endian)
    var result: Long = 0L
    for i <- 0 until byteLength do
      result = result | ((data(offset + 1 + i).toLong & 0xFF) << (8 * i))

    // Mask for top bits from first byte
    val mask = (1 << (8 - byteLength)) - 1
    val topBits = firstByte & mask

    val value = result + (topBits.toLong << (8 * byteLength))
    Some((value, offset + 1 + byteLength))

  /**
   * Simple reader for parsing PVM blobs.
   * Tracks position and uses Varint utilities for varint decoding.
   */
  private class BlobReader(data: Array[Byte]):
    var position: Int = 0

    def remaining: Int = data.length - position

    def readByte(): Option[Int] =
      if position >= data.length then None
      else
        val b = data(position) & 0xFF
        position += 1
        Some(b)

    def readVarintValue(): Option[Int] =
      if position >= data.length then None
      else
        val firstByte = data(position) & 0xFF
        position += 1

        val remainingBytes = data.drop(position)
        Varint.readVarint(remainingBytes, firstByte) match
          case Some((length, value)) =>
            position += length
            Some(value.toInt)
          case None => None

    def readSlice(length: Int): Option[Array[Byte]] =
      if position + length > data.length then None
      else
        val slice = new Array[Byte](length)
        System.arraycopy(data, position, slice, 0, length)
        position += length
        Some(slice)

  /**
   * Parse a PVM blob from raw bytes.
   *
   * The blob format is:
   * - Version byte (0 = 32-bit, 1 = 64-bit)
   * - Varint: ro_data_size
   * - Varint: rw_data_size
   * - Varint: stack_size
   * - Varint: code_size
   * - Varint: jump_table_entry_count
   * - Varint: jump_table_entry_size (if entry_count > 0)
   * - Code section (code_size bytes)
   * - Jump table (entry_count * entry_size bytes)
   * - Bitmask ((code_size + 7) / 8 bytes)
   * - RO data (ro_data_size bytes)
   * - RW data (rw_data_size bytes)
   *
   * @param data The raw blob bytes
   * @return Some(ProgramBlob) on success, None on failure
   */
  def parse(data: Array[Byte]): Option[ProgramBlob] =
    if data.isEmpty then None
    else
      val reader = new BlobReader(data)

      for
        version <- reader.readByte()
        if version <= 1
        is64Bit = version == 1

        roDataSize <- reader.readVarintValue()
        rwDataSize <- reader.readVarintValue()
        stackSize <- reader.readVarintValue()
        codeSize <- reader.readVarintValue()
        if codeSize <= Abi.VmMaximumCodeSize.signed

        jumpTableEntryCount <- reader.readVarintValue()
        jumpTableEntrySize <- if jumpTableEntryCount > 0 then reader.readVarintValue() else Some(0)
        if jumpTableEntryCount <= Abi.VmMaximumJumpTableEntries.signed

        code <- reader.readSlice(codeSize)
        jumpTableSize = jumpTableEntryCount * jumpTableEntrySize
        jumpTableData <- reader.readSlice(jumpTableSize)
        jumpTable = JumpTable(jumpTableData, jumpTableEntrySize)

        bitmaskSize = (codeSize + 7) / 8
        bitmask <- reader.readSlice(bitmaskSize)

        roData <- reader.readSlice(roDataSize)
        rwData <- reader.readSlice(rwDataSize)
      yield ProgramBlob(
        code = code,
        bitmask = bitmask,
        jumpTable = jumpTable,
        is64Bit = is64Bit,
        roData = roData,
        rwData = rwData,
        stackSize = stackSize
      )
